joschahenningsen/TUM-Live
# IPraktikum 73 push notifications

### Motivation and Context
Adds push notification using Firebase Cloud Messaging for whenever a new lecture is uploaded.
Solves https://github.com/TUM-Dev/gocast_mobile/issues/73
### Description
New devices can be registered using the postDeviceToken gRPC method, sending a `device_token` which is then stored for a given user in the `Device` table in the db. Whenever a new lecture is uploaded / a TUM-Live Worker notifies gocast of a finished lecture upload, a fcm notification is then triggered which sends out a push notification to all device tokens owned by the users enrolled in the stream's course. 

#### Changes related to push notifications: 

- [tools/config.go](https://github.com/TUM-Dev/gocast/blob/ddcff6adf5822832073463871d00ae65784071a4/tools/config.go)

- [model/user.go](https://github.com/TUM-Dev/gocast/blob/ddcff6adf5822832073463871d00ae65784071a4/model/user.go)

- [dao/courses.go](https://github.com/TUM-Dev/gocast/blob/ddcff6adf5822832073463871d00ae65784071a4/dao/courses.go)

- [config.yaml](https://github.com/TUM-Dev/gocast/blob/ddcff6adf5822832073463871d00ae65784071a4/config.yaml)

- [api/worker_grpc.go]()

- [cmd/tumlive/tumlive.go](https://github.com/TUM-Dev/gocast/blob/ddcff6adf5822832073463871d00ae65784071a4/cmd/tumlive/tumlive.go)

- [go.mod](https://github.com/TUM-Dev/gocast/blob/ddcff6adf5822832073463871d00ae65784071a4/go.mod)

### Steps for Testing
See the documentation in our [Postman collection](https://www.postman.com/tumdev/workspace/team-workspace/collection/6560b05347e9207e30ae90d3) and this [push notification setup guide](https://gist.github.com/carlobortolan/845141e175c2f43135b3893670f99c59).
